### PR TITLE
Change python build system to be PEP517 compatible

### LINKED ISSUE #2012: Transition from setup tools to the pip install workflow

### PR DESCRIPTION
**Pull request by narendasan**
# Description

The python ecosystem is standardizing on PEP517(https://peps.python.org/pep-0517/) for building packages and while we were kinda compatible it was ultimately hacky. This PR adds support for PEP517 tooling formally.   

Fixes #2012, #2011 

Relevant for #1993 

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -361,15 +367,13 @@ commands:
...
           command: |
-            export CUDA_HOME=/usr/local/cuda-12.1/
+            export CUDA_HOME=/usr/local/cuda-<< parameters.cuda-version >>/
             mv toolchains/ci_workspaces/WORKSPACE.<< parameters.platform >> WORKSPACE
-            cd py
-            python3 -m pip install wheel setuptools
-            python3 -m pip install pybind11==2.6.2
-            python3 setup.py bdist_wheel --fx-only
-            python3 setup.py install --fx-only
+            python3 -m pip install pip==22.0.2
```
Comment (narendasan) [inline .circleci/config.yml:372]: We need to follow up with TensorRT that their python package fails to build with pip versions > 22.x cc @Christina-Young-NVIDIA, @laikhtewari

**Comment by Christina-Young-NVIDIA** [comment]
What is the NVBug number for this?

Thanks, Christina

________________________________

**Comment by narendasan** [comment]
@Christina-Young-NVIDIA we havent opened a bug yet for this, we can discuss at the sync today

File: .circleci/config.yml [hunk 2]
```diff
@@ -816,27 +821,32 @@ jobs:
...
       - unless:
           condition: << parameters.cxx11-abi >>
           steps:
             - when:
                 condition: << parameters.legacy >>
                 steps:
-                  - build-py-legacy
+                  - build-py-legacy:
+                      cuda-version: << parameters.cuda-version >>
+                      torch-build-index: << parameters.torch-build-index >>
```
Comment (gs-olive) [inline .circleci/config.yml:843]: The `legacy` CUDA version and Torch build index will likely differ from the stable version. Consider adding separate parameters for the legacy build (`parameters.cuda-version-legacy`)

File: .circleci/config.yml [hunk 3]
```diff
@@ -1456,6 +1465,7 @@ workflows:
 
       - build-x86_64-linux:
           name: build-x86_64-linux-legacy
+          cuda-version: << pipeline.parameters.cuda-version >>
```
Comment (gs-olive) [inline .circleci/config.yml:1468]: See above regarding legacy CUDA version parameter

File: py/ci/build_whl.sh [hunk 4]
```diff
@@ -8,84 +8,96 @@ export PROJECT_DIR=/workspace/project
...
+    $1/bin/python -m pip install --upgrade pip
+    $1/bin/python -m pip wheel . --config-setting="--build-option=--release" --config-setting="--build-option=--ci" -w dist
+}
+
+patch_wheel() {
+    $2/bin/python -m pip install auditwheel
+    LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$1/torch/lib:$1/tensorrt/:${CUDA_HOME}/lib64:${CUDA_HOME}/lib64/stubs $2/bin/python -m auditwheel repair  $(cat ${PROJECT_DIR}/py/ci/soname_excludes.params) --plat manylinux_2_34_x86_64 dist/torch_tensorrt-*-$3-linux_x86_64.whl
+}
+
 py37() {
```
Comment (gs-olive) [inline py/ci/build_whl.sh:21]: Should the `py37` function be removed since Python 3.7 is not recommended?

File: pyproject.toml [hunk 5]
```diff
@@ -1,25 +1,111 @@
...
+    "Topic :: Software Development :: Libraries",
+]
+readme = {file = "py/README.md", content-type = "text/markdown"}
+requires-python = ">=3.8"
+keywords = ["pytorch", "torch", "tensorrt", "trt", "ai", "artificial intelligence", "ml", "machine learning", "dl", "deep learning", "compiler", "dynamo", "torchscript", "inference"]
+dependencies = [
+    "torch==2.1.0.dev20230619+cu121",
+    "tensorrt>=8.6,<8.7",
+    "packaging>=23",
+    "numpy>=1.24,<1.25",
```
Comment (gs-olive) [inline pyproject.toml:48]: This `numpy` requirement is incompatible with the specification in `requires` above, which has `"numpy>=1.25.0"`. Update for compatibility.

File: pyproject.toml
Comment (gs-olive) [inline pyproject.toml:1]: How would `pyproject.toml` integrate with/support Python-only builds (such as `--fx-only`)?

File: pyproject.toml [hunk 6]
```diff
@@ -1,25 +1,111 @@
...
-    "pyyaml",
-    "setuptools",
-    "cffi",
-    "typing_extensions",
-    "future",
-    "tensorrt >=8.6,<8.7"
+    "setuptools==68.0.0",
+    "packaging>=23.1",
+    "wheel>=0.40.0",
+    "numpy>=1.25.0",
```
Comment (gs-olive) [inline pyproject.toml:6]: This causes an error when building locally - should it be `1.24.0`?

File: pyproject.toml [hunk 7]
```diff
@@ -1,25 +1,111 @@
...
+    "ninja>=1.11.0",
+    "pyyaml>=6.0",
+    "cffi>=1.15.1",
+    "typing_extensions>=4.7.0",
+    "future>=0.18.3",
+    "tensorrt>=8.6,<8.7",
+    #"torch >=2.0.0,<2.1.0",
+    "torch==2.1.0.dev20230619+cu121",
```
Comment (gs-olive) [inline pyproject.toml:14]: This distribution could not be found when building locally - is there a way to specify extra build indices?

**Review by gs-olive — CHANGES_REQUESTED** [review]
Overall, the transformation of `setup.py` looks good. Made suggestions for changes on certain key dependencies and parameters. Also added context and examples from a sample run I did locally using the following command, from the `TensorRT` root directory:
```bash
pip install -e . --use-pep517
```

File: pyproject.toml (hunk 5 above)
Comment (gs-olive) [reply pyproject.toml:48]: See https://github.com/pytorch/TensorRT/pull/2056#discussion_r1251154291, related

File: pyproject.toml (hunk 7 above)
Comment (narendasan) [reply pyproject.toml:14]: You would do this from the commandline

File: .circleci/config.yml (hunk 2 above)
Comment (narendasan) [reply .circleci/config.yml:843]: this would be the case for cudnn too right? I think it should be a separate PR

**Review by gs-olive — COMMENTED** [review]
The changes look great, and it makes the build process much easier with clear error messages for dependencies, build issues, etc. I made some comments regarding versioning.

I successfully built Torch-TRT using this PR, via the following command:
```bash
pip install -e . --use-pep517 --extra-index-url https://download.pytorch.org/whl/nightly/cu121 --extra-index-url https://pypi.ngc.nvidia.com
```

File: py/ci/Dockerfile.ci
Comment (gs-olive) [inline py/ci/Dockerfile.ci:1]: Bazelisk upgrade to 1.17

**Review by gs-olive — CHANGES_REQUESTED** [review]
Consider adding specific build commands to the `README` for ease of use. Additionally, the Docker build system is no longer working, with the following error shown:
```
 /root/.pyenv/versions/3.10.12/bin/python: can't open file '/workspace/torch_tensorrt/src/py/setup.py': [Errno 2] No such file or directory
```
It likely needs some minor updates to `./docker/dist-build.sh`

Other than that, looks good, and the following command works for me. 
```bash
pip install -e . --use-pep517 --extra-index-url https://download.pytorch.org/whl/nightly/cu121 --extra-index-url https://pypi.ngc.nvidia.com/
```

**Comment by narendasan** [comment]
im going to do docs in another pr that has the full docs overhaul

**Comment by narendasan** [comment]
@apbose test this PR in NGC CI

**Pull request merged**